### PR TITLE
 Op-21723:CVE-fixes with java-17 and spring-boot upgrade-v4.1

### PR DESCRIPTION
**Pull request by vivek-opsmx**
https://devopsmx.atlassian.net/browse/OP-21936

OP-21881 involves fixing CVE issues by upgrading to Java 17 and Spring Boot v4.1 .

Since the Fiat service is not utilized in ISD-Argo, I removed the Fiat code from the ISD-Argo Gate Service. However, during this removal process, the /auth/user API stopped returning user details, causing the UI to refresh repeatedly. We resolved this issue by adding "return user" to the code, ensuring that the API now returns the necessary user information without triggering excessive UI refreshing and incorporating a few configurations in the secret settings.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -7,18 +7,33 @@ plugins {
 
 allprojects {
   apply plugin: 'io.spinnaker.project'
+  repositories {
+    mavenLocal()
+    mavenCentral()
```
Comment (rahul-chekuri) [inline build.gradle:12]: where is this method defined?

File: build.gradle [hunk 2]
```diff
@@ -33,43 +48,49 @@ allprojects {
...
+
     configurations.all {
-      exclude group: 'javax.servlet', module: 'servlet-api'
       exclude group: 'javax.servlet', module: 'javax.servlet-api'
-      resolutionStrategy.eachDependency { DependencyResolveDetails details ->
-        if (details.requested.group == 'org.apache.logging.log4j') {
-          details.useVersion '2.19.0'
-        }
-      }
+   //   exclude group: 'javax.servlet', module: 'servlet-api'
```
Comment (rahul-chekuri) [inline build.gradle:72]: remove commented out code

File: docker_build/gate.yml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 services:
   opsmx:
-    baseUrl: http://oes-api:8085
+    baseUrl: http://localhost:8085
```
Comment (rahul-chekuri) [inline docker_build/gate.yml:3]: why has this changed?

File: gate-basic/src/main/java/com/netflix/spinnaker/gate/security/basic/BasicAuthProvider.java [hunk 4]
```diff
@@ -56,42 +68,33 @@ public Authentication authenticate(Authentication authentication) throws Authent
...
+      if (roles != null && !roles.isEmpty() && permissionService != null) {
+        grantedAuthorities.addAll(
+            roles.stream()
+                .map(role -> new SimpleGrantedAuthority(role))
+                .collect(Collectors.toList()));
+        // Updating roles in fiat service
+        permissionService.loginWithRoles(name, roles);
+        log.info("Platform service enabled value :{}", isPlatformEnabled);
+        // Updating roles in platform service
+        if (isPlatformEnabled) {
```
Comment (rahul-chekuri) [inline gate-basic/src/main/java/com/netflix/spinnaker/gate/security/basic/BasicAuthProvider.java:85]: this will never get executed; as the outer if condition will be always false. But we want this block to be executed to support rbac.

Do we use basic auth at any place??

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/filters/FiatSessionFilter.groovy [hunk 5]
```diff
@@ -21,33 +21,32 @@ import com.netflix.spinnaker.fiat.shared.FiatPermissionEvaluator
...
 
 @Slf4j
 class FiatSessionFilter implements Filter {
 
   boolean enabled
-
+  @Value('${services.fiat.enabled:false}')
   FiatStatus fiatStatus
-
+  private List<String> roles;
```
Comment (rahul-chekuri) [inline gate-core/src/main/groovy/com/netflix/spinnaker/gate/filters/FiatSessionFilter.groovy:44]: are we using these roles anywhere?

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/DefaultProviderLookupService.groovy [hunk 6]
```diff
@@ -48,15 +48,14 @@ class DefaultProviderLookupService implements ProviderLookupService, AccountLook
 
   private final AtomicReference<List<AccountDetails>> accountsCache = new AtomicReference<>([])
 
-  @Value('${gate.installation.mode}')
+  @Value('oes')
```
Comment (rahul-chekuri) [inline gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/DefaultProviderLookupService.groovy:51]: is this change intentional?

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/EurekaLookupService.groovy [hunk 7]
```diff
@@ -45,8 +46,12 @@ class EurekaLookupService {
   @Autowired
   ServiceConfiguration serviceConfiguration
 
-  @Autowired
-  OkHttpClient okHttpClient
+  @Bean
+  public OkHttpClient okHttpClient() {
```
Comment (rahul-chekuri) [inline gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/EurekaLookupService.groovy:50]: I don't think this is the right place to add this bean definition.

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/OesAuthorizationService.groovy [hunk 8]
```diff
@@ -13,7 +13,7 @@ import org.springframework.web.bind.annotation.RequestHeader
 import org.springframework.web.bind.annotation.RequestParam
 
 
-@FeignClient(name = "OES", url = '${services.platform.baseUrl}')
+@FeignClient(name = "OES", url = 'http://oes-platform:8095')
 interface OesAuthorizationService {
```
Comment (rahul-chekuri) [inline gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/OesAuthorizationService.groovy:17]: this change should be reverted; we should get URL from config.

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/OesPermissionService.groovy [hunk 9]
```diff
@@ -16,7 +16,7 @@ import static com.netflix.spinnaker.gate.retrofit.UpstreamBadRequest.classifyErr
 @Service
 class OesPermissionService extends PermissionService{
 
-  @Value('${services.platform.enabled}')
+  @Value('true')
```
Comment (rahul-chekuri) [inline gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/OesPermissionService.groovy:19]: this change should be reverted; we should get the value from config.

File: gate-core/src/main/java/com/netflix/spinnaker/gate/services/TokenValidationService.java [hunk 10]
```diff
@@ -7,7 +7,7 @@
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.RequestParam;
 
-@FeignClient(name = "Autopilot", url = "${services.autopilot.baseUrl}")
+@FeignClient(name = "Autopilot", url = "http://oes-autopilot:8090")
 public interface TokenValidationService {
```
Comment (rahul-chekuri) [inline gate-core/src/main/java/com/netflix/spinnaker/gate/services/TokenValidationService.java:11]: this change should be reverted; we should get URL from config.

File: gate-oauth2/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/BearerTokenExtractor.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Copyright 2023 OpsMx, Inc.
```
Comment (rahul-chekuri) [inline gate-oauth2/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/BearerTokenExtractor.java:18]: why 2 copyrights here?

File: gate-oauth2/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/OAuth2SsoConfig.groovy [hunk 12]
```diff
@@ -52,8 +52,8 @@ import javax.servlet.http.HttpServletResponse
 @EnableConfigurationProperties
 // Note the 4 single-quotes below - this is a raw groovy string, because SpEL and groovy
 // string syntax overlap!
-@ConditionalOnExpression('${security.oauth2.enabled:false}')
-class OAuth2SsoConfig extends WebSecurityConfigurerAdapter {
+@ConditionalOnExpression(''''${security.oauth2.client.client-id:}'!=""''')
+class OAuth2SsoConfig {
```
Comment (rahul-chekuri) [inline gate-oauth2/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/OAuth2SsoConfig.groovy:56]: is this expression correct? I doubt it might not work.

File: gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/Saml2UserAttributeMapping.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2024 Netflix, Inc.
+ *
```
Comment (rahul-chekuri) [inline gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/Saml2UserAttributeMapping.java:3]: OpsMx INC

File: gate-web/config/gate.yml [hunk 14]
```diff
@@ -3,6 +3,9 @@ server:
   compression:
     enabled: true
 
+fiat:
+  enabled: true
```
Comment (rahul-chekuri) [inline gate-web/config/gate.yml:7]: why it is true here?

File: gate-web/config/gate.yml [hunk 15]
```diff
@@ -47,6 +50,7 @@ services:
     enabled: false
   fiat:
     enabled: false
+    baseUrl: http://localhost:7003
```
Comment (rahul-chekuri) [inline gate-web/config/gate.yml:53]: why do we need this url?

File: gate-web/gate-web.gradle [hunk 16]
```diff
@@ -25,49 +16,61 @@ dependencies {
...
+  implementation "com.google.guava:guava"
 
-  implementation "com.squareup.okhttp:okhttp"
-  implementation "com.squareup.okhttp:okhttp-urlconnection"
-  implementation "com.squareup.okhttp:okhttp-apache"
-
+  implementation "com.squareup.okhttp:okhttp:2.7.5"
+  implementation "com.squareup.okhttp:okhttp-urlconnection:2.7.5"
+  implementation "com.squareup.okhttp:okhttp-apache:2.7.5"
+  implementation 'org.keycloak:keycloak-admin-client:17.0.0'
```
Comment (rahul-chekuri) [inline gate-web/gate-web.gradle:28]: why keycloak is degraded to 20.0.2?

File: gate-web/src/main/groovy/com/netflix/spinnaker/gate/config/GateConfig.groovy [hunk 17]
```diff
@@ -397,31 +340,23 @@ class GateConfig extends RedisHttpSessionConfiguration {
     }
 
     Endpoint endpoint = serviceConfiguration.getServiceEndpoint(serviceName, dynamicName)
+     log.info(" ******Printing Endpoint value : {}", endpoint )
```
Comment (rahul-chekuri) [inline gate-web/src/main/groovy/com/netflix/spinnaker/gate/config/GateConfig.groovy:343]: this is still needed?

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/service/DatasourceCachingServiceImpl.java [hunk 18]
```diff
@@ -41,7 +41,8 @@ public class DatasourceCachingServiceImpl implements DashboardCachingService {
 
   @Autowired private DatasourceCaching datasourceCaching;
 
-  @Autowired private DashboardClient dashboardClient;
+  @Autowired(required = false)
```
Comment (rahul-chekuri) [inline gate-web/src/main/java/com/opsmx/spinnaker/gate/service/DatasourceCachingServiceImpl.java:44]: remove required attribute

Comment (vivek-opsmx) [reply gate-web/src/main/java/com/opsmx/spinnaker/gate/service/DatasourceCachingServiceImpl.java:44]: Removed

File: build.gradle (hunk 2 above)
Comment (vivek-opsmx) [reply build.gradle:72]: Removed commented code

File: docker_build/gate.yml (hunk 3 above)
Comment (vivek-opsmx) [reply docker_build/gate.yml:3]: for testing purpose i change to localhost Now reverted to Oes-api

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/filters/FiatSessionFilter.groovy (hunk 5 above)
Comment (vivek-opsmx) [reply gate-core/src/main/groovy/com/netflix/spinnaker/gate/filters/FiatSessionFilter.groovy:44]: No not using any where
removed from code

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/DefaultProviderLookupService.groovy (hunk 6 above)
Comment (vivek-opsmx) [reply gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/DefaultProviderLookupService.groovy:51]: reverted

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/OesAuthorizationService.groovy (hunk 8 above)
Comment (vivek-opsmx) [reply gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/OesAuthorizationService.groovy:17]: Reverted

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/OesPermissionService.groovy (hunk 9 above)
Comment (vivek-opsmx) [reply gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/OesPermissionService.groovy:19]: Reverted

File: gate-core/src/main/java/com/netflix/spinnaker/gate/services/TokenValidationService.java (hunk 10 above)
Comment (vivek-opsmx) [reply gate-core/src/main/java/com/netflix/spinnaker/gate/services/TokenValidationService.java:11]: Reverted changes

File: gate-oauth2/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/BearerTokenExtractor.java (hunk 11 above)
Comment (vivek-opsmx) [reply gate-oauth2/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/BearerTokenExtractor.java:18]: Removed

File: gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/Saml2UserAttributeMapping.java (hunk 13 above)
Comment (vivek-opsmx) [reply gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/Saml2UserAttributeMapping.java:3]: Changed to OpsMx INC

File: gate-web/config/gate.yml (hunk 15 above)
Comment (vivek-opsmx) [reply gate-web/config/gate.yml:53]: Removed this URl. for testing purpose i added

File: gate-web/gate-web.gradle (hunk 16 above)
Comment (vivek-opsmx) [reply gate-web/gate-web.gradle:28]: 20.0.2 has some Vulnerabilities now Upgraded to 24.0.4

File: gate-web/src/main/groovy/com/netflix/spinnaker/gate/config/GateConfig.groovy (hunk 17 above)
Comment (vivek-opsmx) [reply gate-web/src/main/groovy/com/netflix/spinnaker/gate/config/GateConfig.groovy:343]: No not needed Removed from code

File: gate-web/src/main/groovy/com/netflix/spinnaker/gate/controllers/OpsmxAutopilotController.groovy [hunk 19]
```diff
@@ -494,26 +490,15 @@ class OpsmxAutopilotController {
     return opsmxAutopilotService.deleteAutoResponse6(type, source, source1, source2, source3, source4)
   }
 
-  @ApiOperation(value = "Endpoint for autopilot rest services")
-  @RequestMapping(value = "/api/{version}/global/template", method = RequestMethod.DELETE)
-  Object deleteAutoResponse7(@PathVariable("version") String version,
-                             @RequestParam(value = "templateType") String templateType,
```
[the remark this answers is not in the document]
Comment (vivek-opsmx) [reply gate-web/src/main/groovy/com/netflix/spinnaker/gate/controllers/OpsmxAutopilotController.groovy:500]: this Api is needed reverted back

File: gate-web/src/main/groovy/com/netflix/spinnaker/gate/controllers/OpsmxDashboardController.groovy [hunk 20]
```diff
@@ -22,45 +22,31 @@ import com.opsmx.spinnaker.gate.factory.dashboard.DashboardCachingServiceBeanFac
...
 
   @Autowired
   OpsmxDashboardService opsmxDashboardService
 
   @Autowired
   DashboardCachingServiceBeanFactory dashboardCachingServiceBeanFactory
 
-  @Value('${gate.installation.mode}')
+  @Value('oes')
   GateInstallationModes gateInstallationMode
```
[the remark this answers is not in the document]
Comment (vivek-opsmx) [reply gate-web/src/main/groovy/com/netflix/spinnaker/gate/controllers/OpsmxDashboardController.groovy:47]: Reverted

File: gate-web/src/main/groovy/com/netflix/spinnaker/gate/services/internal/RoscoServices.groovy [hunk 21]
```diff
@@ -20,7 +20,7 @@ import retrofit.http.GET
 import retrofit.http.Headers
 import retrofit.http.Path
 
-interface RoscoService {
+interface RoscoServices {
```
[the remark this answers is not in the document]
Comment (vivek-opsmx) [reply gate-web/src/main/groovy/com/netflix/spinnaker/gate/services/internal/RoscoServices.groovy:23]: RoscoService Is already present  that's why i changes to RoscoServices

Error: Class 'RoscoService' already exists in 'com.netflix.spinnaker.gate.services.internal'

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/feignclient/AuditService.java [hunk 22]
```diff
@@ -24,7 +24,7 @@
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestHeader;
 
-@FeignClient(name = "auditservice", url = "${services.auditservice.baseUrl}")
+@FeignClient(name = "auditservice", url = "http://oes-audit-service:8097")
```
[the remark this answers is not in the document]
Comment (vivek-opsmx) [reply gate-web/src/main/java/com/opsmx/spinnaker/gate/feignclient/AuditService.java:27]: Reverted

File: gate-web/src/main/java/com/opsmx/spinnaker/gate/feignclient/DashboardClient.java [hunk 23]
```diff
@@ -25,7 +25,7 @@
 import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RequestHeader;
 
-@FeignClient(name = "dashboardservice", url = "${services.dashboard.baseUrl}")
+@FeignClient(name = "dashboardservice", url = " http://oes-dashboard:8094")
```
[the remark this answers is not in the document]
Comment (vivek-opsmx) [reply gate-web/src/main/java/com/opsmx/spinnaker/gate/feignclient/DashboardClient.java:28]: Reverted

File: gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/EurekaLookupService.groovy (hunk 7 above)
Comment (vivek-opsmx) [reply gate-core/src/main/groovy/com/netflix/spinnaker/gate/services/EurekaLookupService.groovy:50]: Correct Added @Autowired

File: gate-basic/src/main/java/com/netflix/spinnaker/gate/security/basic/BasicAuthProvider.java (hunk 4 above)
Comment (vivek-opsmx) [reply gate-basic/src/main/java/com/netflix/spinnaker/gate/security/basic/BasicAuthProvider.java:85]: You Are correct Rahul i only added outer if condition of testing purpose but i did't reverted back.
Now removed outer if condition

File: gate-web/config/gate.yml (hunk 14 above)
Comment (vivek-opsmx) [reply gate-web/config/gate.yml]: Removed "fiat enable =  true"  
earlier i was facing bean issue for i added right now not required

File: build.gradle (hunk 1 above)
Comment (vivek-opsmx) [reply build.gradle]: this changes is needed because Build was failing due to that this code was added by pravallika in V4.0 branch so i added in this also

File: gate-basic/src/main/java/com/netflix/spinnaker/gate/security/basic/BasicAuthProvider.java [hunk 24]
```diff
@@ -56,42 +68,31 @@ public Authentication authenticate(Authentication authentication) throws Authent
...
+              .map(role -> new SimpleGrantedAuthority(role))
+              .collect(Collectors.toList()));
       // Updating roles in fiat service
       permissionService.loginWithRoles(name, roles);
+      log.info("Platform service enabled value :{}", isPlatformEnabled);
       // Updating roles in platform service
-      oesAuthorizationService.cacheUserGroups(roles, name);
+      if (isPlatformEnabled) {
+        oesAuthorizationService.cacheUserGroups(roles, name);
+      }
```
Comment (rahul-chekuri) [inline gate-basic/src/main/java/com/netflix/spinnaker/gate/security/basic/BasicAuthProvider.java:86]: line nos: 73 and 85 both are redundant. Keep one and remove other

File: gate-saml/src/main/groovy/com/opsmx/spinnaker/gate/security/saml/SamlSsoConfig.groovy
Comment (rahul-chekuri) [inline gate-saml/src/main/groovy/com/opsmx/spinnaker/gate/security/saml/SamlSsoConfig.groovy:1]: we dont need this class anymore

File: gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/Saml2UserDetails.java [hunk 25]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 OpsMx, Inc.
```
Comment (rahul-chekuri) [inline gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/Saml2UserDetails.java:2]: year should be 2024

File: gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/SamlSecurityConfiguration.java [hunk 26]
```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2023 OpsMx, Inc.
+ *
```
Comment (rahul-chekuri) [inline gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/SamlSecurityConfiguration.java:3]: same here

File: gate-web/src/main/groovy/com/netflix/spinnaker/gate/services/internal/RoscoServices.groovy (hunk 21 above)
[the remark this answers is not in the document]
Comment (rahul-chekuri) [reply gate-web/src/main/groovy/com/netflix/spinnaker/gate/services/internal/RoscoServices.groovy:23]: argo don't need either of these right

**Review by rahul-chekuri — COMMENTED** [review]
please address the highlighted issue.

File: gate-saml/src/main/groovy/com/opsmx/spinnaker/gate/security/saml/SamlSsoConfig.groovy
Comment (vivek-opsmx) [reply gate-saml/src/main/groovy/com/opsmx/spinnaker/gate/security/saml/SamlSsoConfig.groovy:1]: removed SamlSsoConfig.groovy class

File: gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/Saml2UserDetails.java (hunk 25 above)
Comment (vivek-opsmx) [reply gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/Saml2UserDetails.java:2]: Changed to 2024

File: gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/SamlSecurityConfiguration.java (hunk 26 above)
Comment (vivek-opsmx) [reply gate-saml/src/main/java/com/opsmx/spinnaker/gate/security/saml/SamlSecurityConfiguration.java:3]: Changed to 2024

File: gate-web/src/main/groovy/com/netflix/spinnaker/gate/controllers/OpsmxAuthController.groovy
[the remark this answers is not in the document]
Comment (vivek-opsmx) [reply gate-web/src/main/groovy/com/netflix/spinnaker/gate/controllers/OpsmxAuthController.groovy:1]: Removed Log statement which added in Gate-saml module

File: gate-basic/src/main/java/com/netflix/spinnaker/gate/security/basic/BasicAuthProvider.java (hunk 24 above)
Comment (vivek-opsmx) [reply gate-basic/src/main/java/com/netflix/spinnaker/gate/security/basic/BasicAuthProvider.java:86]: i have only one cacheUserGroups in the code

**Review by rahul-chekuri — APPROVED** [review]
